### ISSUE TITLE
How to get global feature importance for test dataset or new dataset?

### ISSUE THREAD

**Issue by monk1337** [dreamquark-ai/tabnet#358]
Hi, So I am using the tabnet for my project; I have one doubt related to explaining features of tabnet :

After training my model, I get global feature importance for the training dataset; what if I want to test a new dataset and global importance for that particular test dataset? How can I get that?

for train dataset, we can easily get the global importance like this:

`clf.feature_importances_`

for test dataset, I tried something like this:

```
from scipy.sparse import csc_matrix

input_dim      = (df.shape[1] - 1) # excluding target variable
cat_emb_dim    = 1
cat_idxs       = cat_idxs
post_embed_dim = (df.shape[1] - 1) # excluding target variable


explain_matrix, masks  = clf.explain(X_test)
feature_importances_   = explain_matrix.sum(axis=0)
reduce_mat             = create_explain_matrix(input_dim, cat_emb_dim, cat_idxs, post_embed_dim)

feature_importances_   = csc_matrix.dot(feature_importances_,
                                          reduce_mat)
feature_importances_   = feature_importances_ / np.sum(feature_importances_)
```

full example is here:
https://colab.research.google.com/drive/1dgbKXsm94wIECGmqX41Tgv_1qncO7fn1?usp=sharing

Please validate if I am doing it correctly or not.

Thank you!

**Comment by Optimox** [comment]
Hello,

Global feature importance on a test set can be obtained by averaging all the rows of individual explanations.

**Comment by monk1337** [comment]
Hi @Optimox  thanks for reply, I tried that before but the sum of global features are not equal to 1,

if I understood correctly, that's what you suggested:

```
explain_matrix, masks = clf.explain(X_test)
np.mean(explain_matrix, 0)
```

which gives:

```
array([ 14.38177427,  11.21556278, 202.79720104,  28.56499138,
        16.55338611])
```

if I try:

`np.sum(np.mean(explain_matrix, 0))`

it gives 

`273.5129155793108
`

But I am looking for `clf.feature_importances_` where sum of feature importance are 1?

**Comment by Optimox** [comment]
Hello @monk1337,

Yes sorry, the `explain_matrix` is not normalized by default.

So you can try `normalized_individual_imp = explain_matrix / np.sum(explain_matrix, 1)` and then simply take the `np.mean(normalized_individual_imp, 0)` to get global feature importance  on your test set.

**Comment by Optimox** [comment]
@monk1337 is everything good now ?
